examples/bubble_sort.masm

```diff
@@ -0,0 +1,368 @@
+    push.6
+    mem_store.2
+
+    # initialise `g` counter - initialised at 0
+    push.0
+    mem_store.3
+
+    # initialise `adv` length - initialised as the array length using the public inputs
+    mem_store.4
+
+    # Loop over the `advice_stack` pushing values 1-by-1 on the `operand_stack`
+    push.1
+    while.true
+        # Push 1 element from the array into the `operand_stack`
+        adv_push.1
+
+        # Load and increment `g` counter to keep track of added elements
+        mem_load.3
+        add.1
+        dup
+        mem_store.3
+
+        # Load `adv` array length and check if `g` == `adv`;
+        # signifying that all elements from array have been added to `operand_stack`
+        mem_load.4
+        eq
+        if.true
+            # Stop looping and reset `g` to 0
+            push.0
+            push.0
+            mem_store.3
+        else
+            # Continue looping
+            push.1
+        end
+    end
+
+    # Loop over the `operand_stack` pushing values 1-by-1 into memory starting at mem[5]
+    push.1
+    while.true
+        # Offset by 5 considering that we have 5 pre-set variables in `memory`
+        # and that memory is linear meaning that array starts at mem[5]  
+        push.5
+
+        # Load `g` counter and add to 5 enabling incremental insertion
+        mem_load.3
+        add
+
+        # store element at that mem[5 + g]
+        mem_store
+
+        # increment `g` by 1 and overwrite `memory`
+        mem_load.3
+        add.1
+        mem_store.3
+
+        # Load `g` counter and `adv` array length;
+        # check if `g` == `adv; signifying that all elements
+        # have been placed in `memory`
+        mem_load.3
+        mem_load.4
+        eq
+        if.true
+            # Stop looping and reset `g` to 0
+            push.0
+            push.0
+            mem_store.3
+        else
+            # Continue looping
+            push.1
+        end
+    end
+
+    # Prepare the `operand_stack` for sort loop
+    push.1
+end
+
+# swapped() -> void
+#
+# Sets the `swapped` boolean variable to 1 - true
+proc.swapped  
+    push.1
+    mem_store.0
+end
+
+# notSwapped() -> void
+#
+# Sets the `swapped` boolean variable to 0 - false
+#
+proc.notSwapped
+    push.0
+    mem_store.0
+end
+
+# incrementCounters() -> void
+#
+# Increments the `i` and `j` counter variables by 1
+proc.incrementCounters
+    # load `i` counter, increment it by 1 and store it in mem[1]
+    mem_load.1
+    add.1
+    mem_store.1
+
+    # load `j` counter, increment it by 1 and store it in mem[2]
+    mem_load.2
+    add.1
+    mem_store.2
+end
+
+# decrementCounters() -> void
+#
+# Decrements the `i` and `j` counter variables by 1
+proc.decrementCounters
+    # load `i` counter, decrement it by 1 and store it in mem[1]
+    mem_load.1
+    sub.1
+    mem_store.1
+
+    # load `j` counter, decrement by 1 and store it in mem[2]
+    mem_load.2
+    sub.1
+    mem_store.2
+end
```

Comment:
Note: For production code, the `j` would have to be removed. Since it is the same as `i + 1`. All the instructions manipulating it are extra cycle cost that can be removed.